### PR TITLE
[RNMobile] Allow shrink toolbar

### LINKED ISSUE #2180: Allow shrinking the block toolbar for small blocks
It would be good to add an option to stack options in `<BlockMobileToolbar />` into one toolbar button which will open an ActionSheet (ios) / BottomSheet (Android).

Currently, empty `Button` takes up a lot of unnecessary space because the options are spread. That feature allows us to save space and make it look properly.

Current | Desired
--- | ---
Screenshot 2020-04-24 at 09 53 30 | Screenshot 2020-04-24 at 09 52 45

Design:

collapse

More context: https://github.com/WordPress/gutenberg/pull/20191#issuecomment-591637812

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2180

Ref to:
gutenberg-mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2220

That PR introduces the possibility to stack options in <BlockMobileToolbar /> into one toolbar button which will open an ActionSheet (ios) / BottomSheet (Android).

## How has this been tested?

1. Open mobile app
2. Add Buttons block
3. Focus Button block inside Buttons block
4. Expect to see stacked options in the mobile toolbar
5. Press button in the mobile toolbar
6. Expect the disabled button `Move Left` for the first button
7. Expect the disabled button `Move Right` for the last button
8. Try all the options:
	* expect Button is moving left/right when these options are available
	* expect to be able to open settings
	* expect to be able to remove block

## Screenshots 

ios | android
--- | ---
Screenshot 2020-05-05 at 16 57 15 | Screenshot 2020-05-05 at 14 09 44

## Types of changes

* The new type of mobile toolbar named `ShrinkBlockMobileToolbar` along with new prop `allowShrinkToolbar`.

* Add option to disable buttons within Android Picker.

* Two new options to iOS Picker called: `disabledButtonIndices` and `destructiveButtonIndex`

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by lukewalczak** [comment]
Hello @iamthomasbishop, I'm wondering how disabled buttons should look on Android, could you please specify that? Also on the initial mock for Android, there was a separator between `Block Settings` and `Remove Block`, is it valid?

**Comment by iamthomasbishop** [comment]
> I'm wondering how disabled buttons should look on Android

@lukewalczak Ideally, we should hide the button altogether if it's not actionable (at least on Android — that doesn't seem possible on iOS from what I can tell?). If we can't hide it altogether, then it can use a very dim color (very light in light mode and very dark in dark mode). 

We should probably re-use existing styles if possible, but I'm not 100% sure how disabled buttons/icons are styled in code. Based on a quick analysis by pulling screenshots into Figma, it looks like disabled inline toolbar buttons are something like ~30% opacity of our default icon color (or maybe that's a solid color? not sure). 

My suggestion would be to match the icon color with the disabled state of the up/down movers on the inline toolbar. If that is indeed 30% opacity, then maybe we should simply dim the entire cell row to 30% alpha and roll with that. That would look something like this (ignore the icons in this mock):

<a href="https://user-images.githubusercontent.com/1200432/81187122-3c144e80-8f79-11ea-9d35-babdd129074a.png"></a>

> Also on the initial mock for Android, there was a separator between Block Settings and Remove Block, is it valid?

That would be ideal because I think some separation between the primary actions and a destructive action would be nice. If we're going to add a separator, I would add 8px of space above and below the separator. This is how it would look all together:

<a href="https://user-images.githubusercontent.com/1200432/81186996-15561800-8f79-11ea-8766-f064a710d223.png"></a>

Let me know if you have any questions or if that doesn't make sense. 😄

**Comment by iamthomasbishop** [comment]
@lukewalczak Also, one question I have regarding the scope of this feature: Are we making the `•••` toolbar button and menu universal, to all inline toolbars? Or only when it is collapsed because the block is narrow? Maybe a visual will help (forgive the sloppy sketch 😬):

image

With that example in mind, is this ticket aiming to:

A) Collapse into a `•••` menu only when the block is very narrow? Or...
B) Always show the `•••` menu, _and_ collapse the items into the menu when the block is very narrow?

I think the end goal is to have something like B, but I’m not 100% sure if that menu should house _all block actions_, or only the “overflow” ones when the toolbar is collapsed. I would like to at least always have the `Remove block` action in there because this would also help with an issue we’ve been seeing reports of, where users are accidentally deleting blocks too easily, as noted in [this issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1146).

**Comment by iamthomasbishop** [comment]
@lukewalczak I had a chance to review earlier today, this is feeling great already! I jotted down some notes — most of it very minor — some of which is relevant to this ticket, but also another that I will create a ticket for tomorrow.

* As we discussed, the new `ellipsis` (•••) icon from `wordpress/icons` feels too light-weight (each circle is ~2px and the result is a little weak/thin). I've added a _slightly_ heavier version (in horizontal and vertical variations) to the [design docs](https://zeroheight.com/226850145/p/345d49-icons/b/78279c), which feels more appropriate without feeling out of place relative to the other icons. Here's what the icon [looks like in context](https://cloudup.com/cPj7jSE7GTx).
* Can we use the `horizontal` variation of the `•••` icon? It feels a little better for horizontal rhythm and spacing.
* This is counter to my sketch above, but when the block is wide enough to show the `cog`/settings icon, can we give first preference to showing the cog/settings icon before the arrows? So as that space expands, the settings icon would show first, then only when there is enough space to show _both_ mover arrows would we show those.
* Can we change the label of the `Remove Block` action (in the `•••` menu) to `Remove [Block type]`? For example if I go to remove an Image block, the label would become `Remove Image`.
* Not a blocker, but do we have the ability to add icons to the iOS actionSheet? ([example](https://cloudup.com/cCFEaSdsfCV))
* I noticed when I have a single Buttons block that the movers are still part of the ••• menu. Shouldn't they be disabled? ([example](https://cloudup.com/coE9OhWKdRG))

Not relevant to this ticket:

* Specific to the Buttons block: I originally proposed adding a selection border to the selected button — after some time with the button I think we can remove it because the _outer_ selection outline is sufficient ([example](https://cloudup.com/ciCYTnN76ca)) . I can create a ticket for this one tomorrow.

---------

Great work! This is going to be really great to have in the editor — it feels very intuitive and should also solve [this issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1146). // cc @mchowning

**Comment by lukewalczak** [comment]
Hello Thomas! Thanks for the feedback:

> As we discussed, the new ellipsis (•••) icon from wordpress/icons feels too light-weight (each circle is ~2px and the result is a little weak/thin). I've added a slightly heavier version (in horizontal and vertical variations) to the design docs, which feels more appropriate without feeling out of place relative to the other icons. Here's what the icon looks like in context.

Done! I've added `moreHorizontalMobile` icon.

> Can we use the horizontal variation of the ••• icon? It feels a little better for horizontal rhythm and spacing.

Sure!

> This is counter to my sketch above, but when the block is wide enough to show the cog/settings icon, can we give first preference to showing the cog/settings icon before the arrows? So as that space expands, the settings icon would show first, then only when there is enough space to show both mover arrows would we show those.

I've changed the order.

> Can we change the label of the Remove Block action (in the ••• menu) to Remove [Block type]? For example if I go to remove an Image block, the label would become Remove Image.

Done!

> Not a blocker, but do we have the ability to add icons to the iOS actionSheet? (example)

Nope, there is no option for that. This is native ios ActionSheet and it's not supporting that.

> I noticed when I have a single Buttons block that the movers are still part of the ••• menu. Shouldn't they be disabled? (example)

Generally, it's fixed, but I released a testing build from the branch without that change.

File: packages/block-editor/src/components/block-mover/index.native.js [hunk 1]
```diff
@@ -1,52 +1,27 @@
...
-	forwardButtonTitle: __( 'Move block down from row %1$s to row %2$s' ),
-};
+/**
+ * Internal dependencies
+ */
+import {
+	getMoverDescription,
+	getArrowIcon,
+	getMoverButtonTitle,
+} from './mover-description';
```
Comment (jbinda) [inline packages/block-editor/src/components/block-mover/index.native.js:20]: I have prepared some refactor that might looks cleaner.

See [here](https://github.com/WordPress/gutenberg/pull/22217) and let me know WDYT

**Comment by iamthomasbishop** [comment]
Thanks for the update, @lukewalczak  ! Your notes make sense and most everything is working perfectly. Here are some notes based on the new build.

- Empty post: Can we either remove or disable the `•••` button on the Inline Toolbar when the user has an empty post and the first empty paragraph is selected. Note: If we remove, we can probably hide the Inline Toolbar altogether (only on the first block, when the post is empty).
- On the actionSheet, can we change `Block Settings` to `[Block type] Settings`? For example: `Cover Settings` or `Gallery Settings`.

Relevant, but not a blocker:

- Is it possible to make the ActionSheet a Popover on iPad so that the options display directly over the `•••` toggle? ([example](https://cloudup.com/cPilanHPJ9d))

Not specific to this ticket:

- I'm not sure if this is related to this work, but I noticed the "Number of columns" table cell on the Columns block settings looks disabled. ([screenshot](https://cloudup.com/cQi1DXP-D-z))
- Icons look too dim when they are "above" a cover background. ([example](https://cloudup.com/cSL9SJTXlie)). I think this will need some separate time for design consideration.

**Comment by lukewalczak** [comment]
> Empty post: Can we either remove or disable the ••• button on the Inline Toolbar when the user has an empty post and the first empty paragraph is selected. Note: If we remove, we can probably hide the Inline Toolbar altogether (only on the first block, when the post is empty).

That makes sense for me, but I would suggest some improvements to that functionality. First of all, I'm going to keep ••• button and disable `Remove Paragraph` option within Action/Bottom sheet, but only when `Paragraph` is the only block and is empty. 
I wouldn't like to disable ••• button at all, since _soonish_ we are going to add there color options. Also in a case when the user has really long text in `Paragraph` and wants to remove the block, there is no option for that other than removing all characters or adding a new block to remove the previous `Paragraph`. Both of them look wrongly imo. Please let me know wdyt ✌️ 

> On the actionSheet, can we change Block Settings to [Block type] Settings? For example: Cover Settings or Gallery Settings.

Done!

> Is it possible to make the ActionSheet a Popover on iPad so that the options display directly over the ••• toggle? (example)

Done!

> I'm not sure if this is related to this work, but I noticed the "Number of columns" table cell on the Columns block settings looks disabled. (screenshot)

Fixed!

**Comment by iamthomasbishop** [comment]
> First of all, I'm going to keep ••• button and disable Remove Paragraph option within Action/Bottom sheet, but only when Paragraph is the only block and is empty.

Ok, that makes sense 👍 

> I wouldn't like to disable ••• button at all, since soonish we are going to add there color options. 

Good point, I wasn't sure when we'd be adding these but it's good to take into consideration.

Thank you!

**Comment by lukewalczak** [comment]
@iamthomasbishop Currently we have block name near to the action such as `Remove [block name]`, `[block name] Settings`, but for moving block there is a label `Move Block Left/Right`. I think we should pass there block name as well. Wdyt?

**Comment by iamthomasbishop** [comment]
> I think we should pass there block name as well. Wdyt?

@lukewalczak I think we probably could, yea. I'm starting to wonder how blocks with long names will look w/ these labels, esp with long translations. But as long as we're handling them well, it should be okay. Would you mind testing out what a very long block name would do in this scenario — maybe by forcing a long name?

**Comment by lukewalczak** [comment]
Since it's a native component, it's handling that case for us and we don't have to worry about it - message font size will be decreasing until it fits in button.

For the illustration, I forced a _very very_ long name and there are the observations:

regular device font | regular device font | the largest device font
--- | --- | ---
Screenshot 2020-05-14 at 16 14 36 | Screenshot 2020-05-14 at 16 10 05 | Screenshot 2020-05-14 at 16 09 47

**Comment by iamthomasbishop** [comment]
Awesome! Then we're good 😄

File: packages/block-editor/src/components/block-list/block.native.js [hunk 2]
```diff
@@ -95,8 +112,8 @@ class BlockListBlock extends Component {
 
 	render() {
 		const {
-			attributes,
-			blockType,
+			attributes = {},
+			blockType = {},
```
Comment (dratwas) [inline packages/block-editor/src/components/block-list/block.native.js:116]: Is there a possibility to not have the `attributes` and `blockType`? If yes - should we render block with empty object as `attributes` and `blockType` or we can just return `null`?

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+	const options = compact( [
+		wrapBlockMover && backwardButtonOption,
+		wrapBlockMover && forwardButtonOption,
+		wrapBlockSettings && settingsOption,
+		deleteOption,
+	] );
+
+	function onPickerSelect( value ) {
+		if ( value === 'deleteOption' ) {
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:86]: Could you move `deleteOption`, `settingsOption`, `forwardButtonOption`, `backwardButtonOption` to const and use them in options consts and here?

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	const disabledButtonIndices = options
+		.map( ( option, index ) => option.disabled && index + 1 )
+		.filter( Boolean );
+
+	const accessibilityHintIOS = __(
+		'Double tap to open Action Sheet with available options'
+	);
+	const accessibilityHintAndroid = __(
+		'Double tap to open Bottom Sheet with available options'
+	);
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:106]:
```suggestion
	const accessibilityHint = Platform.OS === 'ios' ? __(
		'Double tap to open Action Sheet with available options'
	) :  __(
		'Double tap to open Bottom Sheet with available options'
	);
```

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+	return (
+		<>
+			<ToolbarButton
+				title={ __( 'Open Block Actions Menu' ) }
+				onClick={ () => this.picker.presentPicker() }
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:112]:
```suggestion
				onClick={ this.picker.presentPicker }
```

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js [hunk 6]
```diff
@@ -6,50 +6,60 @@ import { Keyboard, View } from 'react-native';
...
-	order,
 	isStackedHorizontally,
-} ) => (
-	<View style={ styles.toolbar }>
-		<BlockMover
-			clientIds={ [ clientId ] }
-			isStackedHorizontally={ isStackedHorizontally }
-		/>
+	blockWidth,
+	blockMobileToolbarRef,
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/index.native.js:29]: Hmmm, wdyt about a different name for this prop? It's a bit confusing that we pass `blockMobileToolbarRef` to the `block-mobile-toolbar` component

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js [hunk 7]
```diff
@@ -6,50 +6,60 @@ import { Keyboard, View } from 'react-native';
...
 
 /**
  * Internal dependencies
  */
 import styles from './style.scss';
 import BlockMover from '../block-mover';
+import BlockActionsMenu from './block-actions-menu';
 import { BlockSettingsButton } from '../block-settings';
 
+const BREAKPOINTS = {
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/index.native.js:20]: Could you please add a comment about why we need that?

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js [hunk 8]
```diff
@@ -6,50 +6,60 @@ import { Keyboard, View } from 'react-native';
...
+				</BlockSettingsButton.Slot>
 			) }
-			onClick={ onDelete }
-			icon={ trash }
-			extraProps={ { hint: __( 'Double tap to remove the block' ) } }
-		/>
-	</View>
-);
+
+			<BlockActionsMenu
```
Comment (dratwas) [inline packages/block-editor/src/components/block-mobile-toolbar/index.native.js:52]: Do we want to render it even if we don't wrap settings and mover?

File: packages/components/src/mobile/picker/index.android.js [hunk 9]
```diff
@@ -35,30 +46,43 @@ export default class Picker extends Component {
...
-							icon={ option.icon }
-							key={ index }
-							leftAlign={ this.props.leftAlign }
-							label={ option.label }
-							separatorType={ 'none' }
-							onPress={ () => this.onCellPress( option.value ) }
-						/>
+					{ options.map( ( option, index ) => (
+						<>
+							{ options.length > 1 && option.separated && (
```
Comment (dratwas) [inline packages/components/src/mobile/picker/index.android.js:68]: Have you tried to use `separatorType` from the `BottomSheet.Cell`?

File: packages/block-editor/src/components/block-list/block.native.js (hunk 2 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-list/block.native.js:116]: Yup, will do that.

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js (hunk 3 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:86]: I'm going to use `value` directly from option: `value === settingsOption.value`

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js (hunk 4 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:106]: Ok

File: packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js (hunk 5 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:112]: Going to move it to the function with checking if `picker` ref exists.

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js (hunk 7 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js:20]: Will add it.

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js (hunk 8 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js:52]: Yes, we're going to render `BlockActionsMenu` for all the blocks. It's going to keep an option to remove a block for a _full-width_ blocks.

File: packages/components/src/mobile/picker/index.android.js (hunk 9 above)
Comment (lukewalczak) [reply packages/components/src/mobile/picker/index.android.js:68]: In that case we have different requirements for that separator (it has additional vertical padding), so I've decided to create a custom one.

File: packages/block-editor/src/components/block-mobile-toolbar/index.native.js (hunk 6 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js:29]: I think the name is correct and descriptive - the ref is related to `blockMobileToolbar`. Any suggestion how can I rename it?

Comment (dratwas) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js:32]: As far as i can see it is rather a ref to the view that contains the block-mobile-toolbar. I agree that the ref is related to the `blockMobileToolbar` but this is kinda weird that we pass `blockMobileToolbarRef` to the `blockMobileToolbar`, don't you think? The name suggest that this ref points to the `BlockMobileToolbar` component but it doesn't.

Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js]: I've renamed it to `blockMobileToolbarWrapperRef`, however, it looks too long imo.

Comment (dratwas) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js]: Hmmm. yeah, I agree that it's too long. WDYT about something like `anchorNodeRef` or `actionsMenuAnchor`?

Comment (lukewalczak) [reply packages/block-editor/src/components/block-mobile-toolbar/index.native.js]: `anchorNodeRef` makes sense!

**Review by dratwas — APPROVED** [review]
Good job @lukewalczak !

**Review by geriux — COMMENTED** [review]
Hey @lukewalczak 👋 So I left a couple of comments but overall is looking good!

One question, is this expected?:

<kbd></kbd>

I no longer see the trash icon but the "three dots" instead.

**Comment by lukewalczak** [comment]
Hey @geriux, yes it's intended. We are aiming for a new solution to avoid _unexpected_ block [deletion](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1146).

More details in Thomas's [comment](https://github.com/WordPress/gutenberg/pull/22132#issuecomment-624824784)

**Comment by geriux** [comment]
> Hey @geriux, yes it's intended. We are aiming for a new solution to avoid _unexpected_ block [deletion](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1146).
> 
> More details in Thomas's [comment](https://github.com/WordPress/gutenberg/pull/22132#issuecomment-624824784)

Cool! I wanted to double check =) there was quite a lot of messages to keep up. Thanks!

**Review by geriux — APPROVED** [review]
LGTM! Nice work Luke! Tested it again after the changes. 👏

**Pull request merged**